### DISCUSSION TITLE
Use Blitz.js superjson instead of native JSON.serialize/parse

### DISCUSSION THREAD

**Discussion by sergiodxa** [remix-run/remix#1910] [Proposals]
> [!NOTE]
> This is not needed anymore now that Remix uses [turbo-stream](https://github.com/jacob-ebey/turbo-stream) when [Single Fetch](https://remix.run/docs/en/main/guides/single-fetch) feature is enabled, since that allows sending more values than plain JSON, like Date, BigInt, Map, etc.

### What is the new or updated feature that you are suggesting?

Use [superjson](https://github.com/blitz-js/superjson) to serialize and parse data returned from loaders/actions.

### Why should this feature be included?

Right now Remix uses JSON.stringify and JSON.parse, this means types like Date (see #634) are going to be converted to strings when serializing it but not converted back to Date on the revive step, and other types like BigInt are not even compatible with JSON.stringify, so if an ORM returns one the user needs to manually convert them to a string before returning the data/response inside the loader.

This superjson library created by the Blitz.js team have differents stringify and parse method that it's capable of serializing more types of data and reviving them back to the original class (see the table here https://github.com/blitz-js/superjson#parse), this way if a loader does this:

```ts
return json({ date: new Date() })
```

Then the useLoaderData will get this

```ts
let { date } = useLoaderData<{ date: Date }>()
let fullYear = date.getFullYear() // this will work
```

And other native objects like BigInt, RegExp, Set, Map and Error will also work as valid return values.

---

Additionally, superjson can be extended to serialize and revive other objects, it may be cool if Remix let you customize that somehow, but I think just the default config would be better than the native JSON methods.

**Comment by mikestopcontinues** [comment]
It's [`8.7kB` gzipped](https://bundlephobia.com/package/superjson@1.8.0). I don't think it's worth including. It goes against the "web fundamentals" approach of Remix. Better to rally TC39 to update the JSON spec instead.

**Reply by kevlened** [reply]
If you're willing to fudge on the "web fundamentals" a bit, Rich Harris's [devalue](https://github.com/Rich-Harris/devalue) adds 0kB and it's faster. See a comparison [here](https://github.com/Rich-Harris/superjson-and-devalue).

**Reply by thomasfr** [reply]
Out of curiosity, is there a proposal to extend the Spec? Could not find one here: https://github.com/tc39/proposals

**Comment by sergiodxa** [comment]
@mikestopcontinues it's tree-shakeable, so that's not the final size, since the browser only needs one method it should be smaller, but would need to be measured correctly

**Reply by mikestopcontinues** [reply]
For the record, consider me convinced!

**Comment by ghost** [comment]
Could this be supported by having serialize/parse hooks in the client/server entries?

**Comment by sergiodxa** [comment]
> Could this be supported by having serialize/parse hooks in the client/server entries?

I don't think so, you can't change how Remix parse the data from the first request and for requests it does when navigating. This is something that Remix needs to do internally, or somehow expose hooks to do that which is not possible right now.

**Comment by ghost** [comment]
It would be similar to how tRPC does it: https://trpc.io/docs/data-transformers

**Comment by zachdtaylor** [comment]
Does somebody need to give the final word on this before someone can make a PR?

**Comment by sergiodxa** [comment]
@zachtylr21 I already made a PR #662

**Comment by zachdtaylor** [comment]
@sergiodxa in the future, is there a way to know which issues are already being worked on? I wonder if we should use the "assignees" section so people don't accidentally start working on the same thing. Or is it just assumed that if you create the issue, you create the PR as well?

**Comment by sergiodxa** [comment]
@zachtylr21 honestly, I don't know, I also don't know why GitHub didn't related the issue with the PR (I'm sure `closes #` and `fix #` worked for that), I think assignees makes sense but without someone on the team doing it no one can do it.

**Comment by zachdtaylor** [comment]
@sergiodxa Cool, maybe I'll just assume that if someone else created the issue, they're working on it, or I'll just ask them. Btw, I just looked at your PR, I think it's supposed to be "fixes" not "fix".

**Comment by sergiodxa** [comment]
@zachtylr21 just tried with fixes, it doesn't link them, I'm not sure what's happening with GitHub

**Comment by zachdtaylor** [comment]
@sergiodxa weird 🤷

**Comment by MichaelDeBoey** [comment]
> I'm not sure what's happening with GitHub

@sergiodxa I think that's because you're not creating a PR against the default branch (which is `main`)

**Comment by sergiodxa** [comment]
> > I'm not sure what's happening with GitHub
> 
> @sergiodxa I think that's because you're not creating a PR against the default branch (which is `main`)

Ahhh that could be the reason, I think the team should set the default branch as dev so people open PRs automatically against that branch instead of main

**Comment by arthurgeek** [comment]
What is the proposed way to deal with this issue?

Right now, this is what I'm doing, but I'm sure there are better ways (I'm new to TypeScript) :

```typescript
export type TransactionJson = Omit<Transaction, 'createdAt'> & {
  createdAt: string
}

export type LoaderResponse = {
  transactions: TransactionJson[]
}

export const loader: LoaderFunction = async ({ params }) => {
  invariant(params.accountId, 'expected params.accountId')

  const transactions = await getTransactions(params.accountId)

  return json({ transactions })
}

export default function Transactions() {
  const { transactions } = useLoaderData<LoaderResponse>()

  [...]
}
```

Good thing I need an string representation of that Date inside `Transactions`, otherwise I would need to map that `Transaction[]` to parse the `string` as a `Date`.

**Comment by mikestopcontinues** [comment]
@arthurgeek Separate from the `superjson` idea, I've always found it better to store timestamps and convert them to `Date` only when necessary. This was partly for JSON, but mostly just because they cause less headaches. That said, nowadays `date-fns` has made it so I almost never find myself reaching for `Date`.

**Comment by mwarnerdotme** [comment]
As someone who uses Maps and Sets in their (backend) code, I would appreciate support for serializing them.

**Comment by donferi** [comment]
Would love this, currently all my prisma date types are incorrect and it's incredibly frustrating.

**Comment by lucioladen** [comment]
+1
Is kind of annoying this native JSON serialization/parsing lack of support for very quite common types like bigint or dates, specially inuse cases dealing with ORM models (as you might expect, a lot of use cases).
@sergiodxa  did a good job with a lighter-weight `json()` in his remix-utils package, but would be nice for this to be tackled by remix.

**Comment by n1ghtmare** [comment]
Constantly parsing dates is really annoying, this feature would be amazing! Please consider it!

**Comment by gunnartorfis** [comment]
I don't think I can use Remix since I need to serialize BigInt 😔

**Reply by kiliman** [reply]
Just create your own helper functions that use superjson or some other library. I wouldn't throw the baby out with the bath water.

**Reply by gunnartorfis** [reply]
If I only knew how. Would it be enough to return a promise in my loaders that resolves the data in superjson?

**Reply by sergiodxa** [reply]
In Remix Utils I have a useLoaderData/useActionData and json wrappers to let you pass custom parser and serializer functions https://github.com/sergiodxa/remix-utils#useloaderdata

**Reply by ryanwoodcox** [reply]
@kiliman After spending a while combining some custom work with the good work @sergiodxa has done with his remix-utils, I think its worth mentioning I can remove everything I've been customizing in favor of using your wonderful work in https://github.com/kiliman/remix-typedjson. 

(I'm here due because of the Prisma Date serialization typing issue)

**Comment by stephen776** [comment]
Is this still being considered for a future release?  Consider this another vote in favor of this working out of the box 👍

**Reply by kiliman** [reply]
`superjson` is kind of heavy, that's why it's not included in the core. However you can try `remix-typedjson` which is a fast/small (<3k) package for handling native types. I also have `remix-superjson` if you need the full feature set of `superjson`.

**Reply by stephen776** [reply]
Oh yes - I am using `remix-typedjson` already - it's great.

 I was curious if the lighter-weight version would make it to core eventually. 

I believe Next 13 has this sort of end-to-end type handling and with the rise of popularity of things like `tRPC` I figured it would make sense for remix to have similar support out of the box

**Comment by caprica** [comment]
I know some time has passed on this issue, but as someone new to Remix I arrived here from a search after I hit this same issue.

Is there any current recommendation for this date handling in late 2023?

I ask because I checked out remix-typedjson, remix-superjson, superjson-remix, remix-supertyped and the like, and those repos do not seem all that active, with not many recent commits, or what look like on the face of it significant open issues with the current version of Remix.

Situations like this are why it would be very helpful to have a solution for this in Remix core rather than depending on a 3rd-party repo that falls by the wayside as time goes on.

**Reply by sergiodxa** [reply]
It seems unlikely that Remix adds this, Remix-typedjson is probably the most popular of the packages to address this, not many commits doesn’t mean is bad too, the lib is mostly done I believe, although I don’t use it myself

**Reply by lensbart** [reply]
It seems like `remix-typedjson` doesn’t provide types for route `handle` objects, which can depend on `typeof loader`, which may use `typedjson` from `remix-typedjson`. To me, it seems like this kind of stuff would make a lot of sense to be offered out-of-the-box. Do you happen to know why Remix is unlikely to support this? Thanks

**Reply by kiliman** [reply]
I'm not sure what you mean about `handle` objects. They are not serialized so the types will not be modified.

**Reply by sergiodxa** [reply]
I think he meant when you do something like this

```ts
export let handle = {
  something(data: SerializeFrom<typeof loader>) {
    // do something here
  }
}
```

And then you use useMatches to access that handle.something and call it with route.data as argument, similar to how some utils in Remix Utils works, e.g. ExternalScripts or useHydrate.

---

I doubt Remix will add support for superjson or typedjson directly because they want to remain as close to the standards as possible, that's using `response.json`, `JSON.stringify`, and `JSON.parse` instead of custom functions.

What I would do is to pass the data you receive on the `handle` object through superjson or typedjson manually, or maybe wrap this in a function that can do it for you.

**Reply by lensbart** [reply]
Thanks for the pointer, @sergiodxa. The approach you described was the one I tried initially, but I thought I was stuck. Turns out that the `UseDataFunctionReturn` type exposed by `remix-typedjson` allows me to make the types match.

I do feel like, for these cases, it would be helpful if the official Remix docs mentioned the pitfall & pointed the reader to a (list of) package(s) that can help resolve it.

**Reply by sergiodxa** [reply]
Yeah I agree on that, a docs mentioning how serialization will strip some types and how a list of libs you can use to solve it would be useful in the docs.

**Comment by thomasfr** [comment]
I have no preference for or against any existing solution, but would be awesome to have a transparent and automatic way when using `useLoaderData` etc. Remix already does some magic with Promises over the wire when using `defer` in the loader, so the Web Platform argument against this is kinda weak imo.

**Comment by sergiodxa** [comment]
With Single Fetch using turbo-stream, this is not needed anymore since now Remix can send more advanced objects like Date, Map, etc.

**Reply by jackbravo** [reply]
This should be at the top of the thread as well.

**Reply by sergiodxa** [reply]
I will update the original post to mention it
